Do not enable `TLSv1.3` when the current `SSLEngine` does not support it.

For example, when a user is running Armeria on an old JDK without TLSv1.3 support *and* BoringSSL is not loaded, he or she will get the following exception:

```
[armeria-common-worker-epoll-2-1] WARN com.linecorp.armeria.server.HttpServerPipelineConfigurator - [id: 0xc833df45, L:/127.0.0.1:8080 - R:/127.0.0.1:46360] Unexpected exception:
io.netty.handler.codec.DecoderException: java.lang.IllegalArgumentException: TLSv1.3
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:475)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:283)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:154)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:794)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:424)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: TLSv1.3
	at sun.security.ssl.ProtocolVersion.valueOf(ProtocolVersion.java:187)
	at sun.security.ssl.ProtocolList.convert(ProtocolList.java:84)
	at sun.security.ssl.ProtocolList.<init>(ProtocolList.java:52)
	at sun.security.ssl.SSLEngineImpl.setEnabledProtocols(SSLEngineImpl.java:2070)
	at io.netty.handler.ssl.JdkSslContext.configureAndWrapEngine(JdkSslContext.java:341)
	at io.netty.handler.ssl.JdkSslContext.newEngine(JdkSslContext.java:330)
	at io.netty.handler.ssl.SslContext.newHandler(SslContext.java:943)
	at io.netty.handler.ssl.SslContext.newHandler(SslContext.java:935)
	at io.netty.handler.ssl.SniHandler.newSslHandler(SniHandler.java:151)
	at io.netty.handler.ssl.SniHandler.replaceHandler(SniHandler.java:133)
	at io.netty.handler.ssl.SniHandler.onLookupComplete(SniHandler.java:114)
	at io.netty.handler.ssl.AbstractSniHandler.select(AbstractSniHandler.java:211)
	at io.netty.handler.ssl.AbstractSniHandler.decode(AbstractSniHandler.java:203)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:505)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:444)
	... 19 more
[armeria-common-worker-epoll-2-1] WARN com.linecorp.armeria.server.HttpServerPipelineConfigurator - [id: 0xc833df45, L:/127.0.0.1:8080 ! R:/127.0.0.1:46360] Unexpected exception:
io.netty.handler.codec.DecoderException: java.lang.IllegalArgumentException: TLSv1.3
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:475)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:408)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:375)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:358)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:257)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:243)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:236)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1416)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:257)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:243)
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:912)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$8.run(AbstractChannel.java:816)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:416)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:331)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: TLSv1.3
	at sun.security.ssl.ProtocolVersion.valueOf(ProtocolVersion.java:187)
	at sun.security.ssl.ProtocolList.convert(ProtocolList.java:84)
	at sun.security.ssl.ProtocolList.<init>(ProtocolList.java:52)
	at sun.security.ssl.SSLEngineImpl.setEnabledProtocols(SSLEngineImpl.java:2070)
	at io.netty.handler.ssl.JdkSslContext.configureAndWrapEngine(JdkSslContext.java:341)
	at io.netty.handler.ssl.JdkSslContext.newEngine(JdkSslContext.java:330)
	at io.netty.handler.ssl.SslContext.newHandler(SslContext.java:943)
	at io.netty.handler.ssl.SslContext.newHandler(SslContext.java:935)
	at io.netty.handler.ssl.SniHandler.newSslHandler(SniHandler.java:151)
	at io.netty.handler.ssl.SniHandler.replaceHandler(SniHandler.java:133)
	at io.netty.handler.ssl.SniHandler.onLookupComplete(SniHandler.java:114)
	at io.netty.handler.ssl.AbstractSniHandler.select(AbstractSniHandler.java:211)
	at io.netty.handler.ssl.AbstractSniHandler.decode(AbstractSniHandler.java:203)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:505)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:444)
	... 18 more
```

We could avoid enabling `TLSv1.3` when the current `SSLEngine` implementation we are going to use does not support it.

See: https://github.com/line/armeria/blob/44a7e1291c34fa4f6a1f2ae1b876129676b61f19/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java#L66-L78

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -35,7 +35,7 @@ After importing the project, import the IDE settings as well.
 #### IntelliJ IDEA
 
 - [`settings.jar`](https://raw.githubusercontent.com/line/armeria/master/settings/intellij_idea/settings.jar) -
-  See [Importing settings from a JAR archive](https://www.jetbrains.com/help/idea/sharing-your-ide-settings.html#23c8afba).
+  See [Import settings from a ZIP archive](https://www.jetbrains.com/help/idea/sharing-your-ide-settings.html#23c8afba).
 - Make sure to use 'LINE OSS' code style and inspection profile.
   - Go to `Preferences` > `Editors` > `Code Style` and set `Scheme` option to `LINE OSS`.
   - Go to `Preferences` > `Editors` > `Inspections` and set `Profile` option to `LINE OSS`.

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -35,7 +35,7 @@ After importing the project, import the IDE settings as well.
 #### IntelliJ IDEA
 
 - [`settings.jar`](https://raw.githubusercontent.com/line/armeria/master/settings/intellij_idea/settings.jar) -
-  See [Import settings from a ZIP archive](https://www.jetbrains.com/help/idea/sharing-your-ide-settings.html#23c8afba).
+  See [Import settings from a ZIP archive](https://www.jetbrains.com/help/idea/sharing-your-ide-settings.html#7a4f08b8).
 - Make sure to use 'LINE OSS' code style and inspection profile.
   - Go to `Preferences` > `Editors` > `Code Style` and set `Scheme` option to `LINE OSS`.
   - Go to `Preferences` > `Editors` > `Inspections` and set `Profile` option to `LINE OSS`.

--- a/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
@@ -65,8 +65,14 @@ public final class SslContextUtil {
 
     public static final List<String> DEFAULT_PROTOCOLS = ImmutableList.of("TLSv1.3", "TLSv1.2");
 
+    private static final List<String> DEFAULT_JDKENGINE_CIPHERS = ImmutableList.<String>builder()
+            .addAll(Http2SecurityUtil.CIPHERS)
+            .build();
+
+    private static final List<String> DEFAULT_JDKENGINE_PROTOCOLS = ImmutableList.of("TLSv1.2");
+
     /**
-     * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2, TLSv1.3, and
+     * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2, TLSv1.3 (if supported), and
      * TLSv1.2.
      */
     public static SslContext createSslContext(Supplier<SslContextBuilder> sslContextSupplier,
@@ -74,11 +80,16 @@ public final class SslContextUtil {
                                               Consumer<? super SslContextBuilder> userCustomizer) {
         final SslContextBuilder builder = sslContextSupplier.get();
 
-        builder.sslProvider(Flags.useOpenSsl() ? SslProvider.OPENSSL : SslProvider.JDK)
-                  .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]));
-
-        builder.ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
-
+        if (Flags.useOpenSsl()) {
+            builder.sslProvider(SslProvider.OPENSSL)
+                   .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]));
+            builder.ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+        } else {
+            //Netty's JdkSslContext does not support TLSv1.3 by default. In future when it does, we will update.
+            builder.sslProvider(SslProvider.JDK)
+                   .protocols(DEFAULT_JDKENGINE_PROTOCOLS.toArray(new String[0]));
+            builder.ciphers(DEFAULT_JDKENGINE_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+        }
         userCustomizer.accept(builder);
 
         // We called user customization logic before setting ALPN to make sure they don't break

--- a/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
@@ -72,8 +72,8 @@ public final class SslContextUtil {
     private static final List<String> DEFAULT_JDKENGINE_PROTOCOLS = ImmutableList.of("TLSv1.2");
 
     /**
-     * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2, TLSv1.3 (if supported), and
-     * TLSv1.2.
+     * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2,
+     * TLSv1.3 (if supported), and TLSv1.2.
      */
     public static SslContext createSslContext(Supplier<SslContextBuilder> sslContextSupplier,
                                               boolean forceHttp1,

--- a/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
@@ -21,13 +21,18 @@ import java.util.Set;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 
+import javax.net.ssl.SSLEngine;
 import javax.net.ssl.SSLException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 
 import com.linecorp.armeria.common.Flags;
 
+import io.netty.buffer.ByteBufAllocator;
 import io.netty.handler.codec.http2.Http2SecurityUtil;
 import io.netty.handler.ssl.ApplicationProtocolConfig;
 import io.netty.handler.ssl.ApplicationProtocolConfig.Protocol;
@@ -38,12 +43,13 @@ import io.netty.handler.ssl.SslContext;
 import io.netty.handler.ssl.SslContextBuilder;
 import io.netty.handler.ssl.SslProvider;
 import io.netty.handler.ssl.SupportedCipherSuiteFilter;
+import io.netty.util.ReferenceCountUtil;
 
 /**
  * Utilities for configuring {@link SslContextBuilder}.
  */
 public final class SslContextUtil {
-
+    private static final Logger LOGGER = LoggerFactory.getLogger(SslContextUtil.class);
     private static final ApplicationProtocolConfig ALPN_CONFIG = new ApplicationProtocolConfig(
             Protocol.ALPN,
             // NO_ADVERTISE is currently the only mode supported by both OpenSsl and JDK providers.
@@ -82,13 +88,27 @@ public final class SslContextUtil {
 
         if (Flags.useOpenSsl()) {
             builder.sslProvider(SslProvider.OPENSSL)
-                   .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]));
-            builder.ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+                   .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]))
+                   .ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
         } else {
-            //Netty's JdkSslContext does not support TLSv1.3 by default. In future when it does, we will update.
-            builder.sslProvider(SslProvider.JDK)
-                   .protocols(DEFAULT_JDKENGINE_PROTOCOLS.toArray(new String[0]));
-            builder.ciphers(DEFAULT_JDKENGINE_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+            try {
+                // use this to determine actual supported protocols.
+                final SslContextBuilder jdkBuilder = sslContextSupplier.get();
+                final SSLEngine jdkEngine = jdkBuilder.build().newEngine(ByteBufAllocator.DEFAULT);
+                builder.sslProvider(SslProvider.JDK)
+                       .protocols(jdkEngine.getSupportedProtocols())
+                       .ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+                ReferenceCountUtil.release(jdkEngine);
+            } catch (SSLException e) {
+                LOGGER.debug(
+                        "Unable to determine complete list of supported protocol and ciphers. " +
+                        "Will disable TLSv1.3.", e);
+                // Incase of any error, disable TLSv1.3
+                // Netty's JdkSslContext does not support TLSv1.3 by default. Need JDK 11+ for TLS v1.3
+                builder.sslProvider(SslProvider.JDK)
+                       .protocols(DEFAULT_JDKENGINE_PROTOCOLS.toArray(new String[0]))
+                       .ciphers(DEFAULT_JDKENGINE_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+            }
         }
         userCustomizer.accept(builder);
 

--- a/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
@@ -16,6 +16,9 @@
 
 package com.linecorp.armeria.internal;
 
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
 import java.util.List;
 import java.util.Set;
 import java.util.function.Consumer;
@@ -27,12 +30,13 @@ import javax.net.ssl.SSLException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 
 import com.linecorp.armeria.common.Flags;
 
-import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.PooledByteBufAllocator;
 import io.netty.handler.codec.http2.Http2SecurityUtil;
 import io.netty.handler.ssl.ApplicationProtocolConfig;
 import io.netty.handler.ssl.ApplicationProtocolConfig.Protocol;
@@ -49,7 +53,9 @@ import io.netty.util.ReferenceCountUtil;
  * Utilities for configuring {@link SslContextBuilder}.
  */
 public final class SslContextUtil {
-    private static final Logger LOGGER = LoggerFactory.getLogger(SslContextUtil.class);
+
+    private static final Logger logger = LoggerFactory.getLogger(SslContextUtil.class);
+
     private static final ApplicationProtocolConfig ALPN_CONFIG = new ApplicationProtocolConfig(
             Protocol.ALPN,
             // NO_ADVERTISE is currently the only mode supported by both OpenSsl and JDK providers.
@@ -71,11 +77,7 @@ public final class SslContextUtil {
 
     public static final List<String> DEFAULT_PROTOCOLS = ImmutableList.of("TLSv1.3", "TLSv1.2");
 
-    private static final List<String> DEFAULT_JDKENGINE_CIPHERS = ImmutableList.<String>builder()
-            .addAll(Http2SecurityUtil.CIPHERS)
-            .build();
-
-    private static final List<String> DEFAULT_JDKENGINE_PROTOCOLS = ImmutableList.of("TLSv1.2");
+    private static boolean warnedUnsupportedProtocols;
 
     /**
      * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2,
@@ -84,32 +86,32 @@ public final class SslContextUtil {
     public static SslContext createSslContext(Supplier<SslContextBuilder> sslContextSupplier,
                                               boolean forceHttp1,
                                               Consumer<? super SslContextBuilder> userCustomizer) {
-        final SslContextBuilder builder = sslContextSupplier.get();
 
-        if (Flags.useOpenSsl()) {
-            builder.sslProvider(SslProvider.OPENSSL)
-                   .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]))
-                   .ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
-        } else {
-            try {
-                // use this to determine actual supported protocols.
-                final SslContextBuilder jdkBuilder = sslContextSupplier.get();
-                final SSLEngine jdkEngine = jdkBuilder.build().newEngine(ByteBufAllocator.DEFAULT);
-                builder.sslProvider(SslProvider.JDK)
-                       .protocols(jdkEngine.getSupportedProtocols())
-                       .ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
-                ReferenceCountUtil.release(jdkEngine);
-            } catch (SSLException e) {
-                LOGGER.debug(
-                        "Unable to determine complete list of supported protocol and ciphers. " +
-                        "Will disable TLSv1.3.", e);
-                // Incase of any error, disable TLSv1.3
-                // Netty's JdkSslContext does not support TLSv1.3 by default. Need JDK 11+ for TLS v1.3
-                builder.sslProvider(SslProvider.JDK)
-                       .protocols(DEFAULT_JDKENGINE_PROTOCOLS.toArray(new String[0]))
-                       .ciphers(DEFAULT_JDKENGINE_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+        final SslProvider provider = Flags.useOpenSsl() ? SslProvider.OPENSSL : SslProvider.JDK;
+        final SslContextBuilder builder = sslContextSupplier.get();
+        builder.sslProvider(provider);
+
+        final Set<String> supportedProtocols = supportedProtocols(builder);
+        final List<String> protocols = DEFAULT_PROTOCOLS.stream()
+                                                        .filter(supportedProtocols::contains)
+                                                        .collect(toImmutableList());
+        if (protocols.isEmpty()) {
+            throw new IllegalStateException(provider + " supports none of " + DEFAULT_PROTOCOLS);
+        }
+
+        if (!warnedUnsupportedProtocols && DEFAULT_PROTOCOLS.size() != protocols.size()) {
+            warnedUnsupportedProtocols = true;
+            if (logger.isDebugEnabled()) {
+                final List<String> missingProtocols = DEFAULT_PROTOCOLS.stream()
+                                                                       .filter(p -> !protocols.contains(p))
+                                                                       .collect(toImmutableList());
+                logger.debug("{} does not support: {}", provider, missingProtocols);
             }
         }
+
+        builder.protocols(protocols.toArray(new String[0]))
+               .ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+
         userCustomizer.accept(builder);
 
         // We called user customization logic before setting ALPN to make sure they don't break
@@ -118,20 +120,47 @@ public final class SslContextUtil {
             builder.applicationProtocolConfig(ALPN_CONFIG);
         }
 
-        final SslContext sslContext;
+        SslContext sslContext = null;
+        boolean success = false;
         try {
             sslContext = builder.build();
+
+            final Set<String> ciphers = ImmutableSet.copyOf(sslContext.cipherSuites());
+            checkState(!ciphers.isEmpty(), "SSLContext has no cipher suites enabled.");
+
+            if (!forceHttp1) {
+                validateHttp2Ciphers(ciphers);
+            }
+
+            success = true;
+            return sslContext;
         } catch (SSLException e) {
             throw new IllegalStateException("Could not initialize SSL context. Ensure that netty-tcnative is " +
                                             "on the path, this is running on Java 11+, or user customization " +
                                             "of the SSL context is supported by the environment.", e);
+        } finally {
+            if (!success && sslContext != null) {
+                ReferenceCountUtil.release(sslContext);
+            }
         }
+    }
 
-        if (!forceHttp1) {
-            validateHttp2Ciphers(ImmutableSet.copyOf(sslContext.cipherSuites()));
+    @VisibleForTesting
+    static Set<String> supportedProtocols(SslContextBuilder builder) {
+        SslContext ctx = null;
+        SSLEngine engine = null;
+        try {
+            ctx = builder.build();
+            engine = ctx.newEngine(PooledByteBufAllocator.DEFAULT);
+            return ImmutableSet.copyOf(engine.getSupportedProtocols());
+        } catch (Exception e) {
+            throw new IllegalStateException(
+                    "Failed to get the list of supported protocols from an SSLContext.", e);
         }
-
-        return sslContext;
+        finally {
+            ReferenceCountUtil.release(engine);
+            ReferenceCountUtil.release(ctx);
+        }
     }
 
     private static void validateHttp2Ciphers(Set<String> ciphers) {

--- a/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java
@@ -156,8 +156,7 @@ public final class SslContextUtil {
         } catch (Exception e) {
             throw new IllegalStateException(
                     "Failed to get the list of supported protocols from an SSLContext.", e);
-        }
-        finally {
+        } finally {
             ReferenceCountUtil.release(engine);
             ReferenceCountUtil.release(ctx);
         }

--- a/core/src/test/java/com/linecorp/armeria/internal/SslContextUtilTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/SslContextUtilTest.java
@@ -23,6 +23,7 @@ import java.util.Set;
 import org.junit.jupiter.api.Test;
 
 import com.linecorp.armeria.common.Flags;
+import com.linecorp.armeria.common.util.SystemInfo;
 
 import io.netty.handler.ssl.SslContextBuilder;
 import io.netty.handler.ssl.SslProvider;
@@ -41,6 +42,10 @@ class SslContextUtilTest {
     void jdkSsl() {
         final Set<String> supportedProtocols = SslContextUtil.supportedProtocols(
                 SslContextBuilder.forClient().sslProvider(SslProvider.JDK));
-        assertThat(supportedProtocols).contains("TLSv1.2");
+        if (SystemInfo.javaVersion() >= 11) {
+            assertThat(supportedProtocols).contains("TLSv1.2", "TLSv1.3");
+        } else {
+            assertThat(supportedProtocols).contains("TLSv1.2");
+        }
     }
 }

--- a/core/src/test/java/com/linecorp/armeria/internal/SslContextUtilTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/SslContextUtilTest.java
@@ -1,0 +1,46 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.internal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assumptions.assumeThat;
+
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+
+import com.linecorp.armeria.common.Flags;
+
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslProvider;
+
+class SslContextUtilTest {
+
+    @Test
+    void openSsl() {
+        assumeThat(Flags.useOpenSsl()).isTrue();
+        final Set<String> supportedProtocols = SslContextUtil.supportedProtocols(
+                SslContextBuilder.forClient().sslProvider(SslProvider.OPENSSL));
+        assertThat(supportedProtocols).contains("TLSv1.2", "TLSv1.3");
+    }
+
+    @Test
+    void jdkSsl() {
+        final Set<String> supportedProtocols = SslContextUtil.supportedProtocols(
+                SslContextBuilder.forClient().sslProvider(SslProvider.JDK));
+        assertThat(supportedProtocols).contains("TLSv1.2");
+    }
+}

